### ISSUE TITLE
ScatterGL TODO list

### ISSUE THREAD

**Issue by martinRenou** [bloomberg/bqplot#851]
This is related to the work on this PR #661. I'm not sure where to put this, so I'm putting this in an issue.

cc. @maartenbreddels 

- [ ] Allow non-linear color scales. Nothing stops us from making a colormap with non-linear and non-continuous colors in it.
- [ ] Skew is not supported
- [ ] Integer arrays are not supported for `x`, `y`, `color`...
- [ ] Avoid creating a new geometry and associated buffer attributes for every `update_geometry` call. I guess we can create them only once and update their value.
- [ ] clean `update_scene` so that it only renders the scene, material uniforms updates should be in change event callbacks
- [ ] zoom animation not implemented
- [ ] reduce the number of rendering and shader compilation at the creation of the `ScatterGL`

Adding support for missing markers should be fairly easy, missing markers are:
- [ ] `ellipse`: ellipse function already created in fragment shader
- [ ] `triangle-up`: should be easy using the triangle function in fragment shader
- [ ] `triangle-down`: should be easy using the triangle function in fragment shader, inverting the `y` axis
- [ ] `diamond`: maybe we can also reuse the triangle function
- [ ] `rectangle`: rectangle function already there as well

**Comment by martinRenou** [comment]
**Note on the default color:**

When having **3** markers, but a color buffer of size **2**, the `colors` attribute is used a default color in the base scatter plot implementation. The same thing happens when having `NaN`s in the color buffer.
![bqplot_color](https://user-images.githubusercontent.com/21197331/57764372-97638600-7703-11e9-860a-5d44c5997286.png)

In order to achieve the same behavior in the `ScatterGL` implementation, we need to make a few changes. The reason it cannot work right now is that when having a color buffer with a color-scale, we create a colormap texture representing the color-scale, and the color buffer values represent the pixel position where to take the color on the texture.
e.g. the texture would look like:
![texture](https://user-images.githubusercontent.com/21197331/57764975-d3e3b180-7704-11e9-838d-7e3a846aebef.png)

This means that the `color` attribute is of type `float`, this float being the pixel position. But the `colors` values are not part of this colormap so they cannot be represented as `float` number, they would more likely be represented as `vec3` (RGB) in shaders.

**idea1:**
We could fix this by having, in shaders:
```glsl
uniform sampler2D colormap;
uniform vec3 colors[N]; // default colors, N being the max number of values (maximum 200 apparently)

attribute float color; // color position on the colormap texture
attribute float marker_index;
attribute float use_default_color; // if equals to 1.0, use default color

vec3 color_rgb;
if (use_default_color > 0.5) {
    color_rgb = colors[marker_index % N];
} else {
    color_rgb = texture2D(colormap, vec2(color, 0)); // Use texture
}
```
This means adding two attributes. And it does not allow use to have more default colors than the maximum uniform size (around 200 in webgl).

**idea2:**
The second idea would be to put the default_colors in a separate colormap texture.

```glsl
uniform sampler2D colormap;
uniform sampler2D default_colormap; // default colors

attribute float color; // color position on the colormap texture
attribute float use_default_color; // if equals to 1.0, use default color

vec3 color_rgb;
if (use_default_color > 0.5) {
    color_rgb = texture2D(default_colormap, vec2(color, 0)); // Use default texture
} else {
    color_rgb = texture2D(colormap, vec2(color, 0)); // Use colorscale texture
}
```
This means computing the `color` buffer and the `default_colormap` correctly on the CPU before rendering, and it adds one new attribute.
